doc/sslang-lrm.tex

```diff
@@ -328,6 +325,77 @@ \subsection{User-defined operators}
 a look at Agda's user-defined operators.
 
 \section{Types}
+sslang exposes two classes of types. \\
+
+\noindent The first class of types encompasses primitive types: integers, booleans, pure events (i.e. conditional variable), and unit (i.e. void).\\ 
+
+\noindent The second class of types are higher order types that are composed of primitive types:
+\begin{itemize}
+    \item \texttt{[<type>]}: An array containing values of type \texttt{<type>}
+    \item \texttt{(<type>, ...)}: A tuple containing values of type \texttt{<type>}
+    \item \texttt{<type1> $\rightarrow$ <type2>}: A function that takes a   \texttt{<type1>} and returns a  \texttt{<type2>}
+    \item \texttt{\&<type>}: A reference to a variable of type \texttt{<type>}
+    \item \texttt{Sched <type>}: A scheduled variable of type \texttt{<type>}. These variables are scheduled and managed by the SSM runtime and are used to synchronize routines
+\end{itemize}
+
+\noindent This is summarized by the following grammar:
+\setlength{\grammarindent}{5em}
+\begin{grammar}
+<type> ::= [<type>]
+\alt (<type>(, <type>)+)
+\alt <type> $\rightarrow$ <type>
+\alt \&<type>
+\alt Sched <type>
+\alt (<type>)
+\alt Int8 | Int16 | Int32 | Int64
+\alt UInt8 | UInt16 | UInt32 | UInt64
+\alt Bool
+\alt Event
+\alt ()
+\end{grammar}
+
+\section{Routines}
+\noindent Routines are defined at the top-level of a sslang file and consistent of the following components:
+\begin{itemize}
+    \item Routine name
+    \item Zero or more parameters, each of which are optionally typed
+    \item Optionally specified return type
+    \item Routine body
+\end{itemize}
+\noindent The compiler will perform type inference for untyped components and will verify that typed components are correct.
+
+\subsection{Routine Signatures}
+\noindent sslang supports three ways of writing out type signatures for routines.
```

Comment:
I think stating that there are just three ways doesn't account for the variations of functions which take further-nested tuples, i.e., more tuple arguments. For instance, `f ((a: T, b: T), c: T) -> T`, `f (((a, b): (T, T)), c: T) -> T`, `f (((a, b), c): ((T, T), T)) -> T`, and `f ((a, b), c): ((T, T), T)` are all valid way of declaring the equivalent type signature for `f`, and should all be valid in the grammar (though in practice only the first and last are readable).

My suggestion is to base the description off of the following points:
- Our type annotation syntax uses `:` to mean "is of type", and `->` to mean "returns".
- Each function declaration must specify the types of all its arguments, and its return type.
- There are two broad syntactic categories of type signatures: those that look like `f <args> : <fn-type>`, and those that look like `f <typed-args> -> <ret-type>`. The former syntax is evocative of functional languages like OCaml and Haskell, while the latter is evocative of languages like Python and Rust.
- When type-annotating `<typed-args>`, we use `:` to annotate arguments. Our syntax supports "distributing" the `:` over tuple members, so writing `((a, b): (A, B))` is equivalent to `(a: A, b: B)`.
- No identifier should be type-annotated more than once.

This should cover the idea of how to write a syntactically valid type signature for a routine. Those who are interested in working out all the permutations of the above rules may do so while inspecting the provided grammar.

Reply:
> Each function declaration must specify the types of all its arguments, and its return type.

I'm not exactly sure about this, especially with type inference in the pipeline. It was to my understanding that whatever isn't annotated will be filled in and whatever is annotated will be verified by the type checker. As such, the grammar makes type annotations optional.

Reply:
We haven't worked out the details or extent of the type system just yet. Depending on what flavor of type inference we do, we may be forced to annotate the type (for some type systems, esp those involving type classes, inference is undecidable). We can easily relax this requirement once that's worked out. 

Also, even if we do have inference capabilities, I still think it's a good idea to annotate types for top-level functions. Even if they're not needed by the type checker, they perform an important role of conveying the interface of a function to programmers.

Reply:
The issue is the two forms of tuple annotation. I made the type annotations optional because it simplifies the grammar so much when dealing with tuples. The idea there was that a subsequent pass would verify existence of annotations and only one annotation per variable. I can change it to be mandatory though. Please let me know.